{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!jupyter nbconvert --to script 'topicModelParameterTuning.ipynb'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.3.2\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "from pyspark.ml.linalg import Vectors, SparseVector\n",
    "from pyspark.ml.clustering import LDA\n",
    "from pyspark.sql import SQLContext\n",
    "import re\n",
    "from nltk.corpus import stopwords\n",
    "from pyspark import SparkContext\n",
    "\n",
    "sc = SparkContext.getOrCreate()\n",
    "# check if spark context is defined\n",
    "print(sc.version)\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = sc.wholeTextFiles(\"file:///data2/sunandan/sparkex/sshrotri/allAbstracts/*.txt\").map(lambda x: x[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tokenize the data\n",
    "tokens = data.map(lambda document: document.strip().lower()) \\\n",
    "    .map(lambda x: x.split()) \\\n",
    "    .map(lambda word: [x for x in word if x.isalpha()]) \\\n",
    "    .map(lambda word: [x for x in word if len(x) > 2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['cytochrome',\n",
       "  'enzyme',\n",
       "  'metabolizes',\n",
       "  'the',\n",
       "  'anticancer',\n",
       "  'prodrug',\n",
       "  'cyclophosphamide',\n",
       "  'which',\n",
       "  'decomposes',\n",
       "  'the',\n",
       "  'cytotoxic',\n",
       "  'metabolites',\n",
       "  'acrolein',\n",
       "  'and',\n",
       "  'phosphoramide',\n",
       "  'have',\n",
       "  'evaluated',\n",
       "  'the',\n",
       "  'bystander',\n",
       "  'cytotoxicity',\n",
       "  'cpa',\n",
       "  'combination',\n",
       "  'with',\n",
       "  'enzyme',\n",
       "  'prodrug',\n",
       "  'therapy',\n",
       "  'using',\n",
       "  'cell',\n",
       "  'agarose',\n",
       "  'overlay',\n",
       "  'this',\n",
       "  'method',\n",
       "  'mimics',\n",
       "  'the',\n",
       "  'tumor',\n",
       "  'microenvironment',\n",
       "  'limiting',\n",
       "  'the',\n",
       "  'diffusion',\n",
       "  'metabolites',\n",
       "  'and',\n",
       "  'reducing',\n",
       "  'the',\n",
       "  'oxygen',\n",
       "  'concentration',\n",
       "  'levels',\n",
       "  'similar',\n",
       "  'those',\n",
       "  'found',\n",
       "  'solid',\n",
       "  'under',\n",
       "  'these',\n",
       "  'the',\n",
       "  'cyp',\n",
       "  'activity',\n",
       "  'tumor',\n",
       "  'cells',\n",
       "  'was',\n",
       "  'decreased',\n",
       "  'compared',\n",
       "  'standard',\n",
       "  'aerobic',\n",
       "  'despite',\n",
       "  'this',\n",
       "  'decrease',\n",
       "  'metabolic',\n",
       "  'potent',\n",
       "  'bystander',\n",
       "  'effect',\n",
       "  'was',\n",
       "  'resulting',\n",
       "  'killing',\n",
       "  'cpa',\n",
       "  'tumor',\n",
       "  'cell',\n",
       "  'population',\n",
       "  'comprised',\n",
       "  'only',\n",
       "  'tumor',\n",
       "  'transient',\n",
       "  'transfection',\n",
       "  'small',\n",
       "  'fraction',\n",
       "  'human',\n",
       "  'hepatoma',\n",
       "  'cell',\n",
       "  'population',\n",
       "  'with',\n",
       "  'expression',\n",
       "  'plasmid',\n",
       "  'followed',\n",
       "  'treatment',\n",
       "  'with',\n",
       "  'cpa',\n",
       "  'led',\n",
       "  'eradication',\n",
       "  'the',\n",
       "  'such',\n",
       "  'bystander',\n",
       "  'effect',\n",
       "  'was',\n",
       "  'observed',\n",
       "  'without',\n",
       "  'the',\n",
       "  'agarose',\n",
       "  'these',\n",
       "  'findings',\n",
       "  'suggest',\n",
       "  'that',\n",
       "  'the',\n",
       "  'agarose',\n",
       "  'overlay',\n",
       "  'technique',\n",
       "  'very',\n",
       "  'useful',\n",
       "  'vitro',\n",
       "  'test',\n",
       "  'system',\n",
       "  'for',\n",
       "  'investigation',\n",
       "  'the',\n",
       "  'bystander',\n",
       "  'effect',\n",
       "  'and',\n",
       "  'other',\n",
       "  'combinations',\n",
       "  'under',\n",
       "  'conditions',\n",
       "  'that',\n",
       "  'mimic',\n",
       "  'the',\n",
       "  'hypoxic',\n",
       "  'conditions',\n",
       "  'present',\n",
       "  'solid',\n",
       "  'tumors'],\n",
       " ['this',\n",
       "  'review',\n",
       "  'summarizes',\n",
       "  'clinical',\n",
       "  'experience',\n",
       "  'with',\n",
       "  'infectious',\n",
       "  'disease',\n",
       "  'vaccines',\n",
       "  'and',\n",
       "  'data',\n",
       "  'from',\n",
       "  'animal',\n",
       "  'tumor',\n",
       "  'models',\n",
       "  'that',\n",
       "  'support',\n",
       "  'paradigm',\n",
       "  'shift',\n",
       "  'for',\n",
       "  'cancer',\n",
       "  'vaccines',\n",
       "  'from',\n",
       "  'therapeutic',\n",
       "  'prevention'],\n",
       " ['because',\n",
       "  'their',\n",
       "  'ability',\n",
       "  'the',\n",
       "  'relationships',\n",
       "  'oncolytic',\n",
       "  'viruses',\n",
       "  'cannot',\n",
       "  'easily',\n",
       "  'better',\n",
       "  'understand',\n",
       "  'the',\n",
       "  'pharmacokinetics',\n",
       "  'virotherapy',\n",
       "  'relation',\n",
       "  'viral',\n",
       "  'dose',\n",
       "  'and',\n",
       "  'administered',\n",
       "  'intraperitoneally',\n",
       "  'orthotopic',\n",
       "  'mouse',\n",
       "  'model',\n",
       "  'ovarian',\n",
       "  'attenuated',\n",
       "  'oncolytic',\n",
       "  'measles',\n",
       "  'virus',\n",
       "  'engineered',\n",
       "  'express',\n",
       "  'soluble',\n",
       "  'human',\n",
       "  'carcinoembryonic',\n",
       "  'antigen',\n",
       "  'and',\n",
       "  'the',\n",
       "  'virus',\n",
       "  'currently',\n",
       "  'undergoing',\n",
       "  'phase',\n",
       "  'clinical',\n",
       "  'testing',\n",
       "  'patients',\n",
       "  'with',\n",
       "  'ovarian',\n",
       "  'plasma',\n",
       "  'cea',\n",
       "  'levels',\n",
       "  'correlate',\n",
       "  'with',\n",
       "  'numbers',\n",
       "  'tumor',\n",
       "  'cells',\n",
       "  'given',\n",
       "  'and',\n",
       "  'were',\n",
       "  'used',\n",
       "  'surrogate',\n",
       "  'monitor',\n",
       "  'the',\n",
       "  'profiles',\n",
       "  'viral',\n",
       "  'gene',\n",
       "  'expression',\n",
       "  'over',\n",
       "  'the',\n",
       "  'antineoplastic',\n",
       "  'activity',\n",
       "  'was',\n",
       "  'apparent',\n",
       "  'over',\n",
       "  'wide',\n",
       "  'range',\n",
       "  'virus',\n",
       "  'doses',\n",
       "  'with',\n",
       "  'little',\n",
       "  'reduction',\n",
       "  'observed',\n",
       "  'antitumor',\n",
       "  'even',\n",
       "  'the',\n",
       "  'lowest',\n",
       "  'tested',\n",
       "  'analysis',\n",
       "  'cea',\n",
       "  'profiles',\n",
       "  'treated',\n",
       "  'mice',\n",
       "  'was',\n",
       "  'highly',\n",
       "  'illustrating',\n",
       "  'the',\n",
       "  'variability',\n",
       "  'virus',\n",
       "  'kinetics',\n",
       "  'different',\n",
       "  'dose',\n",
       "  'the',\n",
       "  'highest',\n",
       "  'doses',\n",
       "  'virus',\n",
       "  'were',\n",
       "  'associated',\n",
       "  'with',\n",
       "  'higher',\n",
       "  'initial',\n",
       "  'levels',\n",
       "  'tumor',\n",
       "  'cell',\n",
       "  'but',\n",
       "  'the',\n",
       "  'final',\n",
       "  'outcome',\n",
       "  'therapy',\n",
       "  'all',\n",
       "  'dose',\n",
       "  'levels',\n",
       "  'was',\n",
       "  'partial',\n",
       "  'equilibrium',\n",
       "  'between',\n",
       "  'virus',\n",
       "  'and',\n",
       "  'resulting',\n",
       "  'significant',\n",
       "  'slowing',\n",
       "  'tumor',\n",
       "  'growth',\n",
       "  'and',\n",
       "  'enhanced',\n",
       "  'survival',\n",
       "  'the'],\n",
       " ['mortality',\n",
       "  'due',\n",
       "  'colorectal',\n",
       "  'cancer',\n",
       "  'high',\n",
       "  'and',\n",
       "  'associated',\n",
       "  'with',\n",
       "  'the',\n",
       "  'development',\n",
       "  'liver',\n",
       "  'approximately',\n",
       "  'human',\n",
       "  'crcs',\n",
       "  'harbor',\n",
       "  'activating',\n",
       "  'mutation',\n",
       "  'the',\n",
       "  'kras',\n",
       "  'tumor',\n",
       "  'cells',\n",
       "  'with',\n",
       "  'activated',\n",
       "  'kras',\n",
       "  'are',\n",
       "  'particularly',\n",
       "  'sensitive',\n",
       "  'reovirus',\n",
       "  'oncolytic',\n",
       "  'the',\n",
       "  'efficacy',\n",
       "  'therapies',\n",
       "  'may',\n",
       "  'positively',\n",
       "  'negatively',\n",
       "  'modulated',\n",
       "  'the',\n",
       "  'host',\n",
       "  'immune',\n",
       "  'this',\n",
       "  'study',\n",
       "  'was',\n",
       "  'designed',\n",
       "  'assess',\n",
       "  'the',\n",
       "  'effect',\n",
       "  'immunosuppression',\n",
       "  'reovirus',\n",
       "  'oncolysis',\n",
       "  'established',\n",
       "  'colorectal',\n",
       "  'micrometastases',\n",
       "  'the',\n",
       "  'mouse',\n",
       "  'crc',\n",
       "  'cells',\n",
       "  'harbor',\n",
       "  'mutant',\n",
       "  'kras',\n",
       "  'gene',\n",
       "  'and',\n",
       "  'are',\n",
       "  'susceptible',\n",
       "  'oncolysis',\n",
       "  'reovirus',\n",
       "  'isolated',\n",
       "  'liver',\n",
       "  'tumors',\n",
       "  'were',\n",
       "  'established',\n",
       "  'syngenic',\n",
       "  'immunocompetent',\n",
       "  'mice',\n",
       "  'intrahepatic',\n",
       "  'reovirus',\n",
       "  'therapy',\n",
       "  'was',\n",
       "  'given',\n",
       "  'single',\n",
       "  'intratumoral',\n",
       "  'injection',\n",
       "  'control',\n",
       "  'mice',\n",
       "  'and',\n",
       "  'cyclosporin',\n",
       "  'immunosuppressed',\n",
       "  'tumor',\n",
       "  'growth',\n",
       "  'was',\n",
       "  'analyzed',\n",
       "  'over',\n",
       "  'time',\n",
       "  'bioluminescence',\n",
       "  'the',\n",
       "  'outgrowth',\n",
       "  'established',\n",
       "  'crc',\n",
       "  'liver',\n",
       "  'metastases',\n",
       "  'immunocompetent',\n",
       "  'mice',\n",
       "  'was',\n",
       "  'efficiently',\n",
       "  'but',\n",
       "  'temporarily',\n",
       "  'inhibited',\n",
       "  'with',\n",
       "  'single',\n",
       "  'injection',\n",
       "  'reovirus',\n",
       "  'immunosuppression',\n",
       "  'with',\n",
       "  'cyclosporin',\n",
       "  'markedly',\n",
       "  'increased',\n",
       "  'and',\n",
       "  'prolonged',\n",
       "  'the',\n",
       "  'therapeutic',\n",
       "  'effect',\n",
       "  'and',\n",
       "  'allowed',\n",
       "  'complete',\n",
       "  'reovirus',\n",
       "  'tumor',\n",
       "  'eradication',\n",
       "  'subpopulation',\n",
       "  'the',\n",
       "  'conclude',\n",
       "  'that',\n",
       "  'reovirus',\n",
       "  'effective',\n",
       "  'therapeutic',\n",
       "  'agent',\n",
       "  'against',\n",
       "  'established',\n",
       "  'colorectal',\n",
       "  'liver',\n",
       "  'metastases',\n",
       "  'and',\n",
       "  'that',\n",
       "  'immunosuppression',\n",
       "  'enhances',\n",
       "  'treatment'],\n",
       " ['previous',\n",
       "  'studies',\n",
       "  'have',\n",
       "  'shown',\n",
       "  'that',\n",
       "  'the',\n",
       "  'human',\n",
       "  'melanoma',\n",
       "  'has',\n",
       "  'activity',\n",
       "  'vitro',\n",
       "  'and',\n",
       "  'vitro',\n",
       "  'studies',\n",
       "  'using',\n",
       "  'human',\n",
       "  'peripheral',\n",
       "  'blood',\n",
       "  'mononuclear',\n",
       "  'cells',\n",
       "  'indicate',\n",
       "  'that',\n",
       "  'has',\n",
       "  'the',\n",
       "  'individual',\n",
       "  'properties',\n",
       "  'have',\n",
       "  'been',\n",
       "  'previously',\n",
       "  'examined',\n",
       "  'there',\n",
       "  'not',\n",
       "  'single',\n",
       "  'study',\n",
       "  'that',\n",
       "  'has',\n",
       "  'examined',\n",
       "  'antitumor',\n",
       "  'and',\n",
       "  'proimmune',\n",
       "  'properties',\n",
       "  'the',\n",
       "  'tumor',\n",
       "  'suppressive',\n",
       "  'activity',\n",
       "  'and',\n",
       "  'the',\n",
       "  'cytokine',\n",
       "  'activity',\n",
       "  'have',\n",
       "  'not',\n",
       "  'been',\n",
       "  'previously',\n",
       "  'tested',\n",
       "  'immunocompetent',\n",
       "  'therefore',\n",
       "  'the',\n",
       "  'present',\n",
       "  'study',\n",
       "  'evaluated',\n",
       "  'the',\n",
       "  'antitumor',\n",
       "  'and',\n",
       "  'immune',\n",
       "  'properties',\n",
       "  'murine',\n",
       "  'syngeneic',\n",
       "  'tumor',\n",
       "  'gene',\n",
       "  'transfer',\n",
       "  'murine',\n",
       "  'fibrosarcoma',\n",
       "  'and',\n",
       "  'normal',\n",
       "  'cells',\n",
       "  'significantly',\n",
       "  'inhibited',\n",
       "  'growth',\n",
       "  'and',\n",
       "  'induced',\n",
       "  'apoptosis',\n",
       "  'tumor',\n",
       "  'cells',\n",
       "  'but',\n",
       "  'not',\n",
       "  'normal',\n",
       "  'intratumoral',\n",
       "  'administration',\n",
       "  'resulted',\n",
       "  'significant',\n",
       "  'inhibition',\n",
       "  'tumor',\n",
       "  'growth',\n",
       "  'with',\n",
       "  'subset',\n",
       "  'mice',\n",
       "  'showing',\n",
       "  'complete',\n",
       "  'tumor',\n",
       "  'next',\n",
       "  'evaluated',\n",
       "  'the',\n",
       "  'immune',\n",
       "  'potentiation',\n",
       "  'activity',\n",
       "  'cancer',\n",
       "  'vaccine',\n",
       "  'cells',\n",
       "  'transfected',\n",
       "  'with',\n",
       "  'and',\n",
       "  'injected',\n",
       "  'into',\n",
       "  'syngeneic',\n",
       "  'immunocompetent',\n",
       "  'mice',\n",
       "  'were',\n",
       "  'unable',\n",
       "  'they',\n",
       "  'did',\n",
       "  'grow',\n",
       "  'immunocompromised',\n",
       "  'nude',\n",
       "  'these',\n",
       "  'when',\n",
       "  'challenged',\n",
       "  'with',\n",
       "  'parental',\n",
       "  'tumor',\n",
       "  'cells',\n",
       "  'experienced',\n",
       "  'tumor',\n",
       "  'suggesting',\n",
       "  'induction',\n",
       "  'systemic',\n",
       "  'splenocytes',\n",
       "  'prepared',\n",
       "  'from',\n",
       "  'vaccinated',\n",
       "  'mice',\n",
       "  'demonstrated',\n",
       "  'higher',\n",
       "  'proliferative',\n",
       "  'activity',\n",
       "  'and',\n",
       "  'produced',\n",
       "  'elevated',\n",
       "  'levels',\n",
       "  'cytokines',\n",
       "  'compared',\n",
       "  'with',\n",
       "  'those',\n",
       "  'from',\n",
       "  'control',\n",
       "  'vitro',\n",
       "  'subset',\n",
       "  'analysis',\n",
       "  'splenocytes',\n",
       "  'from',\n",
       "  'vaccinated',\n",
       "  'mice',\n",
       "  'demonstrated',\n",
       "  'significant',\n",
       "  'increase',\n",
       "  'the',\n",
       "  'but',\n",
       "  'not',\n",
       "  'the',\n",
       "  'cell',\n",
       "  'population',\n",
       "  'thus',\n",
       "  'treatment',\n",
       "  'syngeneic',\n",
       "  'tumors',\n",
       "  'induces',\n",
       "  'tumor',\n",
       "  'cell',\n",
       "  'death',\n",
       "  'and',\n",
       "  'promotes',\n",
       "  'immune',\n",
       "  'leading',\n",
       "  'anticancer']]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tokens.take(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "stop = stopwords.words('english')\n",
    "# Get vocabulary \n",
    "# Flat map to put words into one giant list\n",
    "# Remove stopwords\n",
    "# Map and then reduce by key\n",
    "# Sort by word count\n",
    "termCounts = tokens.flatMap(lambda document: document) \\\n",
    "    .filter(lambda x: x not in stop) \\\n",
    "    .map(lambda word: (word,1)) \\\n",
    "    .reduceByKey(lambda x,y: x+y) \\\n",
    "    .map(lambda tuple: (tuple[1],tuple[0])) \\\n",
    "    .sortByKey(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(12096, 'gene'),\n",
       " (11168, 'cells'),\n",
       " (9428, 'expression'),\n",
       " (9028, 'cell'),\n",
       " (9028, 'cancer'),\n",
       " (7785, 'tumor'),\n",
       " (7289, 'patients'),\n",
       " (6981, 'genetic'),\n",
       " (5893, 'human'),\n",
       " (5285, 'associated')]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "termCounts.take(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionary\n",
    "vocabulary = termCounts.map(lambda x: x[1]) \\\n",
    "    .zipWithIndex() \\\n",
    "    .collectAsMap()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gene': 0,\n",
       " 'cells': 1,\n",
       " 'expression': 2,\n",
       " 'cell': 3,\n",
       " 'cancer': 4,\n",
       " 'tumor': 5,\n",
       " 'patients': 6,\n",
       " 'genetic': 7,\n",
       " 'human': 8,\n",
       " 'associated': 9,\n",
       " 'clinical': 10,\n",
       " 'results': 11,\n",
       " 'genes': 12,\n",
       " 'using': 13,\n",
       " 'study': 14,\n",
       " 'may': 15,\n",
       " 'also': 16,\n",
       " 'protein': 17,\n",
       " 'two': 18,\n",
       " 'identified': 19,\n",
       " 'variants': 20,\n",
       " 'mutations': 21,\n",
       " 'analysis': 22,\n",
       " 'data': 23,\n",
       " 'found': 24,\n",
       " 'disease': 25,\n",
       " 'association': 26,\n",
       " 'increased': 27,\n",
       " 'growth': 28,\n",
       " 'showed': 29,\n",
       " 'significantly': 30,\n",
       " 'novel': 31,\n",
       " 'role': 32,\n",
       " 'dna': 33,\n",
       " 'treatment': 34,\n",
       " 'therapy': 35,\n",
       " 'studies': 36,\n",
       " 'one': 37,\n",
       " 'compared': 38,\n",
       " 'levels': 39,\n",
       " 'significant': 40,\n",
       " 'risk': 41,\n",
       " 'mice': 42,\n",
       " 'therapeutic': 43,\n",
       " 'potential': 44,\n",
       " 'used': 45,\n",
       " 'mutation': 46,\n",
       " 'including': 47,\n",
       " 'breast': 48,\n",
       " 'could': 49,\n",
       " 'high': 50,\n",
       " 'observed': 51,\n",
       " 'factor': 52,\n",
       " 'effect': 53,\n",
       " 'activity': 54,\n",
       " 'three': 55,\n",
       " 'sequencing': 56,\n",
       " 'vector': 57,\n",
       " 'show': 58,\n",
       " 'tumors': 59,\n",
       " 'effects': 60,\n",
       " 'new': 61,\n",
       " 'different': 62,\n",
       " 'suggest': 63,\n",
       " 'findings': 64,\n",
       " 'development': 65,\n",
       " 'target': 66,\n",
       " 'important': 67,\n",
       " 'molecular': 68,\n",
       " 'revealed': 69,\n",
       " 'family': 70,\n",
       " 'demonstrated': 71,\n",
       " 'common': 72,\n",
       " 'signaling': 73,\n",
       " 'reported': 74,\n",
       " 'immune': 75,\n",
       " 'type': 76,\n",
       " 'promoter': 77,\n",
       " 'variant': 78,\n",
       " 'well': 79,\n",
       " 'vectors': 80,\n",
       " 'vivo': 81,\n",
       " 'previously': 82,\n",
       " 'control': 83,\n",
       " 'virus': 84,\n",
       " 'individuals': 85,\n",
       " 'among': 86,\n",
       " 'response': 87,\n",
       " 'receptor': 88,\n",
       " 'activation': 89,\n",
       " 'evidence': 90,\n",
       " 'polymorphisms': 91,\n",
       " 'vitro': 92,\n",
       " 'model': 93,\n",
       " 'first': 94,\n",
       " 'reduced': 95,\n",
       " 'number': 96,\n",
       " 'multiple': 97,\n",
       " 'region': 98,\n",
       " 'genomic': 99,\n",
       " 'within': 100,\n",
       " 'although': 101,\n",
       " 'mouse': 102,\n",
       " 'higher': 103,\n",
       " 'functional': 104,\n",
       " 'present': 105,\n",
       " 'lung': 106,\n",
       " 'testing': 107,\n",
       " 'several': 108,\n",
       " 'delivery': 109,\n",
       " 'provide': 110,\n",
       " 'known': 111,\n",
       " 'patient': 112,\n",
       " 'cases': 113,\n",
       " 'complex': 114,\n",
       " 'specific': 115,\n",
       " 'allele': 116,\n",
       " 'survival': 117,\n",
       " 'inhibition': 118,\n",
       " 'performed': 119,\n",
       " 'normal': 120,\n",
       " 'primary': 121,\n",
       " 'syndrome': 122,\n",
       " 'induced': 123,\n",
       " 'susceptibility': 124,\n",
       " 'expressed': 125,\n",
       " 'identify': 126,\n",
       " 'function': 127,\n",
       " 'major': 128,\n",
       " 'screening': 129,\n",
       " 'report': 130,\n",
       " 'demonstrate': 131,\n",
       " 'involved': 132,\n",
       " 'approach': 133,\n",
       " 'population': 134,\n",
       " 'increase': 135,\n",
       " 'use': 136,\n",
       " 'highly': 137,\n",
       " 'viral': 138,\n",
       " 'single': 139,\n",
       " 'whether': 140,\n",
       " 'transcription': 141,\n",
       " 'enhanced': 142,\n",
       " 'loss': 143,\n",
       " 'shown': 144,\n",
       " 'based': 145,\n",
       " 'kinase': 146,\n",
       " 'pathway': 147,\n",
       " 'four': 148,\n",
       " 'factors': 149,\n",
       " 'whereas': 150,\n",
       " 'decreased': 151,\n",
       " 'transfer': 152,\n",
       " 'investigated': 153,\n",
       " 'effective': 154,\n",
       " 'mechanisms': 155,\n",
       " 'combination': 156,\n",
       " 'lines': 157,\n",
       " 'detected': 158,\n",
       " 'apoptosis': 159,\n",
       " 'many': 160,\n",
       " 'cellular': 161,\n",
       " 'rare': 162,\n",
       " 'large': 163,\n",
       " 'research': 164,\n",
       " 'affected': 165,\n",
       " 'targeting': 166,\n",
       " 'rna': 167,\n",
       " 'phenotype': 168,\n",
       " 'chromosome': 169,\n",
       " 'via': 170,\n",
       " 'system': 171,\n",
       " 'mechanism': 172,\n",
       " 'snps': 173,\n",
       " 'resulted': 174,\n",
       " 'either': 175,\n",
       " 'low': 176,\n",
       " 'developed': 177,\n",
       " 'resistance': 178,\n",
       " 'test': 179,\n",
       " 'prostate': 180,\n",
       " 'level': 181,\n",
       " 'caused': 182,\n",
       " 'without': 183,\n",
       " 'group': 184,\n",
       " 'early': 185,\n",
       " 'transduction': 186,\n",
       " 'cause': 187,\n",
       " 'overexpression': 188,\n",
       " 'families': 189,\n",
       " 'targeted': 190,\n",
       " 'proliferation': 191,\n",
       " 'information': 192,\n",
       " 'suggesting': 193,\n",
       " 'characterized': 194,\n",
       " 'polymorphism': 195,\n",
       " 'severe': 196,\n",
       " 'antitumor': 197,\n",
       " 'small': 198,\n",
       " 'total': 199,\n",
       " 'efficacy': 200,\n",
       " 'frequency': 201,\n",
       " 'linkage': 202,\n",
       " 'stem': 203,\n",
       " 'deletion': 204,\n",
       " 'presence': 205,\n",
       " 'due': 206,\n",
       " 'mrna': 207,\n",
       " 'expressing': 208,\n",
       " 'binding': 209,\n",
       " 'confirmed': 210,\n",
       " 'combined': 211,\n",
       " 'tissue': 212,\n",
       " 'diagnosis': 213,\n",
       " 'blood': 214,\n",
       " 'strategy': 215,\n",
       " 'following': 216,\n",
       " 'controls': 217,\n",
       " 'sequence': 218,\n",
       " 'candidate': 219,\n",
       " 'liver': 220,\n",
       " 'med': 221,\n",
       " 'similar': 222,\n",
       " 'genet': 223,\n",
       " 'recent': 224,\n",
       " 'proteins': 225,\n",
       " 'likely': 226,\n",
       " 'pathogenic': 227,\n",
       " 'indicate': 228,\n",
       " 'various': 229,\n",
       " 'encoding': 230,\n",
       " 'key': 231,\n",
       " 'samples': 232,\n",
       " 'loci': 233,\n",
       " 'lower': 234,\n",
       " 'analyzed': 235,\n",
       " 'injection': 236,\n",
       " 'infection': 237,\n",
       " 'transcriptional': 238,\n",
       " 'metastasis': 239,\n",
       " 'critical': 240,\n",
       " 'current': 241,\n",
       " 'alleles': 242,\n",
       " 'additional': 243,\n",
       " 'mutant': 244,\n",
       " 'health': 245,\n",
       " 'metastatic': 246,\n",
       " 'regulation': 247,\n",
       " 'genome': 248,\n",
       " 'responses': 249,\n",
       " 'inhibited': 250,\n",
       " 'treated': 251,\n",
       " 'haplotype': 252,\n",
       " 'analyses': 253,\n",
       " 'progression': 254,\n",
       " 'adenovirus': 255,\n",
       " 'evaluated': 256,\n",
       " 'examined': 257,\n",
       " 'carcinoma': 258,\n",
       " 'populations': 259,\n",
       " 'ability': 260,\n",
       " 'systemic': 261,\n",
       " 'ovarian': 262,\n",
       " 'time': 263,\n",
       " 'regions': 264,\n",
       " 'transgene': 265,\n",
       " 'related': 266,\n",
       " 'might': 267,\n",
       " 'efficient': 268,\n",
       " 'murine': 269,\n",
       " 'oncolytic': 270,\n",
       " 'diagnostic': 271,\n",
       " 'disorder': 272,\n",
       " 'recombinant': 273,\n",
       " 'genotype': 274,\n",
       " 'direct': 275,\n",
       " 'age': 276,\n",
       " 'mitochondrial': 277,\n",
       " 'tested': 278,\n",
       " 'regulatory': 279,\n",
       " 'resulting': 280,\n",
       " 'correlated': 281,\n",
       " 'understanding': 282,\n",
       " 'autosomal': 283,\n",
       " 'strong': 284,\n",
       " 'five': 285,\n",
       " 'administration': 286,\n",
       " 'nuclear': 287,\n",
       " 'cohort': 288,\n",
       " 'knockdown': 289,\n",
       " 'medical': 290,\n",
       " 'models': 291,\n",
       " 'variation': 292,\n",
       " 'determine': 293,\n",
       " 'positive': 294,\n",
       " 'production': 295,\n",
       " 'independent': 296,\n",
       " 'leading': 297,\n",
       " 'support': 298,\n",
       " 'result': 299,\n",
       " 'disorders': 300,\n",
       " 'induction': 301,\n",
       " 'remains': 302,\n",
       " 'formation': 303,\n",
       " 'established': 304,\n",
       " 'death': 305,\n",
       " 'bone': 306,\n",
       " 'methylation': 307,\n",
       " 'would': 308,\n",
       " 'features': 309,\n",
       " 'genetics': 310,\n",
       " 'required': 311,\n",
       " 'underlying': 312,\n",
       " 'diseases': 313,\n",
       " 'thus': 314,\n",
       " 'possible': 315,\n",
       " 'healthy': 316,\n",
       " 'previous': 317,\n",
       " 'locus': 318,\n",
       " 'recessive': 319,\n",
       " 'epithelial': 320,\n",
       " 'melanoma': 321,\n",
       " 'changes': 322,\n",
       " 'associations': 323,\n",
       " 'pathways': 324,\n",
       " 'pancreatic': 325,\n",
       " 'heterozygous': 326,\n",
       " 'homozygous': 327,\n",
       " 'rate': 328,\n",
       " 'interaction': 329,\n",
       " 'method': 330,\n",
       " 'promising': 331,\n",
       " 'investigate': 332,\n",
       " 'induce': 333,\n",
       " 'brain': 334,\n",
       " 'years': 335,\n",
       " 'improved': 336,\n",
       " 'form': 337,\n",
       " 'invasion': 338,\n",
       " 'replication': 339,\n",
       " 'provides': 340,\n",
       " 'drug': 341,\n",
       " 'subjects': 342,\n",
       " 'snp': 343,\n",
       " 'ratio': 344,\n",
       " 'containing': 345,\n",
       " 'differences': 346,\n",
       " 'reduction': 347,\n",
       " 'clinically': 348,\n",
       " 'poor': 349,\n",
       " 'improve': 350,\n",
       " 'malignant': 351,\n",
       " 'japanese': 352,\n",
       " 'inflammatory': 353,\n",
       " 'children': 354,\n",
       " 'exon': 355,\n",
       " 'markers': 356,\n",
       " 'missense': 357,\n",
       " 'adenoviral': 358,\n",
       " 'oncogenic': 359,\n",
       " 'muscle': 360,\n",
       " 'tissues': 361,\n",
       " 'inhibitor': 362,\n",
       " 'led': 363,\n",
       " 'migration': 364,\n",
       " 'contribute': 365,\n",
       " 'six': 366,\n",
       " 'serum': 367,\n",
       " 'individual': 368,\n",
       " 'impact': 369,\n",
       " 'plasmid': 370,\n",
       " 'efficiency': 371,\n",
       " 'limited': 372,\n",
       " 'across': 373,\n",
       " 'strategies': 374,\n",
       " 'less': 375,\n",
       " 'suggested': 376,\n",
       " 'least': 377,\n",
       " 'transduced': 378,\n",
       " 'sensitivity': 379,\n",
       " 'nucleotide': 380,\n",
       " 'often': 381,\n",
       " 'colorectal': 382,\n",
       " 'carrying': 383,\n",
       " 'activated': 384,\n",
       " 'inherited': 385,\n",
       " 'included': 386,\n",
       " 'determined': 387,\n",
       " 'available': 388,\n",
       " 'indicated': 389,\n",
       " 'types': 390,\n",
       " 'despite': 391,\n",
       " 'suggests': 392,\n",
       " 'phosphorylation': 393,\n",
       " 'review': 394,\n",
       " 'influence': 395,\n",
       " 'recently': 396,\n",
       " 'participants': 397,\n",
       " 'genetically': 398,\n",
       " 'negative': 399,\n",
       " 'identification': 400,\n",
       " 'consistent': 401,\n",
       " 'essential': 402,\n",
       " 'antigen': 403,\n",
       " 'domain': 404,\n",
       " 'directly': 405,\n",
       " 'site': 406,\n",
       " 'colon': 407,\n",
       " 'carriers': 408,\n",
       " 'distinct': 409,\n",
       " 'quantitative': 410,\n",
       " 'describe': 411,\n",
       " 'carrier': 412,\n",
       " 'endothelial': 413,\n",
       " 'located': 414,\n",
       " 'assessed': 415,\n",
       " 'relevant': 416,\n",
       " 'groups': 417,\n",
       " 'transfection': 418,\n",
       " 'targets': 419,\n",
       " 'conducted': 420,\n",
       " 'evaluate': 421,\n",
       " 'considered': 422,\n",
       " 'even': 423,\n",
       " 'described': 424,\n",
       " 'generated': 425,\n",
       " 'spectrum': 426,\n",
       " 'tumour': 427,\n",
       " 'acid': 428,\n",
       " 'european': 429,\n",
       " 'approaches': 430,\n",
       " 'elevated': 431,\n",
       " 'central': 432,\n",
       " 'short': 433,\n",
       " 'enzyme': 434,\n",
       " 'autoimmune': 435,\n",
       " 'phenotypes': 436,\n",
       " 'develop': 437,\n",
       " 'obtained': 438,\n",
       " 'familial': 439,\n",
       " 'biological': 440,\n",
       " 'exome': 441,\n",
       " 'studied': 442,\n",
       " 'tests': 443,\n",
       " 'functions': 444,\n",
       " 'dominant': 445,\n",
       " 'promotes': 446,\n",
       " 'derived': 447,\n",
       " 'metabolic': 448,\n",
       " 'deficiency': 449,\n",
       " 'useful': 450,\n",
       " 'haplotypes': 451,\n",
       " 'upon': 452,\n",
       " 'implicated': 453,\n",
       " 'need': 454,\n",
       " 'roles': 455,\n",
       " 'class': 456,\n",
       " 'future': 457,\n",
       " 'potentially': 458,\n",
       " 'followed': 459,\n",
       " 'taken': 460,\n",
       " 'copy': 461,\n",
       " 'cancers': 462,\n",
       " 'basis': 463,\n",
       " 'responsible': 464,\n",
       " 'increasing': 465,\n",
       " 'history': 466,\n",
       " 'process': 467,\n",
       " 'strongly': 468,\n",
       " 'diabetes': 469,\n",
       " 'women': 470,\n",
       " 'aim': 471,\n",
       " 'cytokine': 472,\n",
       " 'suppressor': 473,\n",
       " 'methods': 474,\n",
       " 'linked': 475,\n",
       " 'upregulated': 476,\n",
       " 'overall': 477,\n",
       " 'phenotypic': 478,\n",
       " 'indicating': 479,\n",
       " 'chronic': 480,\n",
       " 'congenital': 481,\n",
       " 'greater': 482,\n",
       " 'sites': 483,\n",
       " 'promote': 484,\n",
       " 'injected': 485,\n",
       " 'silencing': 486,\n",
       " 'glioma': 487,\n",
       " 'novo': 488,\n",
       " 'mediated': 489,\n",
       " 'line': 490,\n",
       " 'frequently': 491,\n",
       " 'inhibitors': 492,\n",
       " 'set': 493,\n",
       " 'lead': 494,\n",
       " 'unrelated': 495,\n",
       " 'active': 496,\n",
       " 'xenograft': 497,\n",
       " 'correlation': 498,\n",
       " 'days': 499,\n",
       " 'coding': 500,\n",
       " 'prevalence': 501,\n",
       " 'peripheral': 502,\n",
       " 'affect': 503,\n",
       " 'genotyped': 504,\n",
       " 'potent': 505,\n",
       " 'assess': 506,\n",
       " 'antibody': 507,\n",
       " 'downstream': 508,\n",
       " 'enhance': 509,\n",
       " 'cardiac': 510,\n",
       " 'alternative': 511,\n",
       " 'able': 512,\n",
       " 'cytotoxic': 513,\n",
       " 'vascular': 514,\n",
       " 'assay': 515,\n",
       " 'cycle': 516,\n",
       " 'regulated': 517,\n",
       " 'still': 518,\n",
       " 'frequencies': 519,\n",
       " 'detection': 520,\n",
       " 'decrease': 521,\n",
       " 'better': 522,\n",
       " 'express': 523,\n",
       " 'gastric': 524,\n",
       " 'deletions': 525,\n",
       " 'suppressed': 526,\n",
       " 'hcc': 527,\n",
       " 'unique': 528,\n",
       " 'conditions': 529,\n",
       " 'selected': 530,\n",
       " 'increases': 531,\n",
       " 'transfected': 532,\n",
       " 'majority': 533,\n",
       " 'complete': 534,\n",
       " 'marker': 535,\n",
       " 'crc': 536,\n",
       " 'per': 537,\n",
       " 'regulator': 538,\n",
       " 'developmental': 539,\n",
       " 'regression': 540,\n",
       " 'leads': 541,\n",
       " 'members': 542,\n",
       " 'chromosomal': 543,\n",
       " 'confidence': 544,\n",
       " 'provided': 545,\n",
       " 'suppression': 546,\n",
       " 'genotypes': 547,\n",
       " 'reporter': 548,\n",
       " 'receptors': 549,\n",
       " 'case': 550,\n",
       " 'aav': 551,\n",
       " 'upregulation': 552,\n",
       " 'isolated': 553,\n",
       " 'range': 554,\n",
       " 'importance': 555,\n",
       " 'pathogenesis': 556,\n",
       " 'specifically': 557,\n",
       " 'regarding': 558,\n",
       " 'host': 559,\n",
       " 'therapies': 560,\n",
       " 'sample': 561,\n",
       " 'remain': 562,\n",
       " 'addition': 563,\n",
       " 'local': 564,\n",
       " 'maternal': 565,\n",
       " 'knowledge': 566,\n",
       " 'intellectual': 567,\n",
       " 'vaccine': 568,\n",
       " 'lentiviral': 569,\n",
       " 'exhibited': 570,\n",
       " 'significance': 571,\n",
       " 'applied': 572,\n",
       " 'lack': 573,\n",
       " 'causes': 574,\n",
       " 'especially': 575,\n",
       " 'distribution': 576,\n",
       " 'variations': 577,\n",
       " 'signal': 578,\n",
       " 'particularly': 579,\n",
       " 'infected': 580,\n",
       " 'produced': 581,\n",
       " 'repair': 582,\n",
       " 'aimed': 583,\n",
       " 'differentiation': 584,\n",
       " 'chain': 585,\n",
       " 'absence': 586,\n",
       " 'hearing': 587,\n",
       " 'inhibit': 588,\n",
       " 'second': 589,\n",
       " 'downregulation': 590,\n",
       " 'nude': 591,\n",
       " 'surface': 592,\n",
       " 'prenatal': 593,\n",
       " 'predicted': 594,\n",
       " 'animal': 595,\n",
       " 'animals': 596,\n",
       " 'invasive': 597,\n",
       " 'intratumoral': 598,\n",
       " 'adult': 599,\n",
       " 'male': 600,\n",
       " 'depletion': 601,\n",
       " 'involving': 602,\n",
       " 'relative': 603,\n",
       " 'size': 604,\n",
       " 'eight': 605,\n",
       " 'seven': 606,\n",
       " 'general': 607,\n",
       " 'subsequent': 608,\n",
       " 'rates': 609,\n",
       " 'developing': 610,\n",
       " 'currently': 611,\n",
       " 'regulates': 612,\n",
       " 'splicing': 613,\n",
       " 'onset': 614,\n",
       " 'therefore': 615,\n",
       " 'chinese': 616,\n",
       " 'damage': 617,\n",
       " 'sequences': 618,\n",
       " 'part': 619,\n",
       " 'sle': 620,\n",
       " 'induces': 621,\n",
       " 'interactions': 622,\n",
       " 'selection': 623,\n",
       " 'plasma': 624,\n",
       " 'yet': 625,\n",
       " 'accumulation': 626,\n",
       " 'efficiently': 627,\n",
       " 'seen': 628,\n",
       " 'molecules': 629,\n",
       " 'application': 630,\n",
       " 'fusion': 631,\n",
       " 'sirna': 632,\n",
       " 'assays': 633,\n",
       " 'newborn': 634,\n",
       " 'presented': 635,\n",
       " 'panel': 636,\n",
       " 'represent': 637,\n",
       " 'array': 638,\n",
       " 'experimental': 639,\n",
       " 'defects': 640,\n",
       " 'include': 641,\n",
       " 'dose': 642,\n",
       " 'extracellular': 643,\n",
       " 'tool': 644,\n",
       " 'heart': 645,\n",
       " 'repeat': 646,\n",
       " 'herpes': 647,\n",
       " 'status': 648,\n",
       " 'estimated': 649,\n",
       " 'resistant': 650,\n",
       " 'rat': 651,\n",
       " 'long': 652,\n",
       " 'modified': 653,\n",
       " 'leukemia': 654,\n",
       " 'renal': 655,\n",
       " 'effectively': 656,\n",
       " 'egfr': 657,\n",
       " 'natural': 658,\n",
       " 'advanced': 659,\n",
       " 'interval': 660,\n",
       " 'given': 661,\n",
       " 'represents': 662,\n",
       " 'affecting': 663,\n",
       " 'crucial': 664,\n",
       " 'hereditary': 665,\n",
       " 'variety': 666,\n",
       " 'simplex': 667,\n",
       " 'genotyping': 668,\n",
       " 'received': 669,\n",
       " 'work': 670,\n",
       " 'parental': 671,\n",
       " 'skin': 672,\n",
       " 'endogenous': 673,\n",
       " 'frequent': 674,\n",
       " 'epigenetic': 675,\n",
       " 'diagnosed': 676,\n",
       " 'thereby': 677,\n",
       " 'utility': 678,\n",
       " 'capacity': 679,\n",
       " 'largely': 680,\n",
       " 'parents': 681,\n",
       " 'conventional': 682,\n",
       " 'alone': 683,\n",
       " 'achieved': 684,\n",
       " 'transition': 685,\n",
       " 'specificity': 686,\n",
       " 'female': 687,\n",
       " 'regulating': 688,\n",
       " 'fibroblasts': 689,\n",
       " 'regulate': 690,\n",
       " 'odds': 691,\n",
       " 'acute': 692,\n",
       " 'hla': 693,\n",
       " 'care': 694,\n",
       " 'alterations': 695,\n",
       " 'recurrent': 696,\n",
       " 'luciferase': 697,\n",
       " 'help': 698,\n",
       " 'viruses': 699,\n",
       " 'order': 700,\n",
       " 'promoted': 701,\n",
       " 'measured': 702,\n",
       " 'toward': 703,\n",
       " 'protective': 704,\n",
       " 'stable': 705,\n",
       " 'relatively': 706,\n",
       " 'membrane': 707,\n",
       " 'chromatin': 708,\n",
       " 'carried': 709,\n",
       " 'phase': 710,\n",
       " 'conclude': 711,\n",
       " 'generation': 712,\n",
       " 'matrix': 713,\n",
       " 'main': 714,\n",
       " 'western': 715,\n",
       " 'value': 716,\n",
       " 'much': 717,\n",
       " 'constructed': 718,\n",
       " 'heterogeneity': 719,\n",
       " 'disequilibrium': 720,\n",
       " 'detect': 721,\n",
       " 'difference': 722,\n",
       " 'variable': 723,\n",
       " 'evaluation': 724,\n",
       " 'explore': 725,\n",
       " 'selective': 726,\n",
       " 'secondary': 727,\n",
       " 'dependent': 728,\n",
       " 'implications': 729,\n",
       " 'propose': 730,\n",
       " 'cnvs': 731,\n",
       " 'designed': 732,\n",
       " 'impaired': 733,\n",
       " 'cytotoxicity': 734,\n",
       " 'necrosis': 735,\n",
       " 'shows': 736,\n",
       " 'characteristics': 737,\n",
       " 'aggressive': 738,\n",
       " 'inhibits': 739,\n",
       " 'weeks': 740,\n",
       " 'plays': 741,\n",
       " 'little': 742,\n",
       " 'play': 743,\n",
       " 'aberrant': 744,\n",
       " 'public': 745,\n",
       " 'african': 746,\n",
       " 'experiments': 747,\n",
       " 'intracellular': 748,\n",
       " 'together': 749,\n",
       " 'providing': 750,\n",
       " 'fluorescent': 751,\n",
       " 'reveal': 752,\n",
       " 'severity': 753,\n",
       " 'fetal': 754,\n",
       " 'solid': 755,\n",
       " 'subset': 756,\n",
       " 'component': 757,\n",
       " 'overexpressed': 758,\n",
       " 'heterogeneous': 759,\n",
       " 'prognosis': 760,\n",
       " 'safety': 761,\n",
       " 'abnormal': 762,\n",
       " 'outcomes': 763,\n",
       " 'comprehensive': 764,\n",
       " 'germline': 765,\n",
       " 'skeletal': 766,\n",
       " 'approximately': 767,\n",
       " 'antibodies': 768,\n",
       " 'expansion': 769,\n",
       " 'encodes': 770,\n",
       " 'mammary': 771,\n",
       " 'engineered': 772,\n",
       " 'microarray': 773,\n",
       " 'stress': 774,\n",
       " 'initial': 775,\n",
       " 'comparison': 776,\n",
       " 'causative': 777,\n",
       " 'defined': 778,\n",
       " 'poorly': 779,\n",
       " 'immunity': 780,\n",
       " 'sufficient': 781,\n",
       " 'somatic': 782,\n",
       " 'emt': 783,\n",
       " 'mass': 784,\n",
       " 'causal': 785,\n",
       " 'delivered': 786,\n",
       " 'abnormalities': 787,\n",
       " 'score': 788,\n",
       " 'histone': 789,\n",
       " 'anticancer': 790,\n",
       " 'ligand': 791,\n",
       " 'angiogenesis': 792,\n",
       " 'altered': 793,\n",
       " 'contributes': 794,\n",
       " 'outcome': 795,\n",
       " 'pattern': 796,\n",
       " 'reduce': 797,\n",
       " 'retinal': 798,\n",
       " 'structure': 799,\n",
       " 'proportion': 800,\n",
       " 'involvement': 801,\n",
       " 'quality': 802,\n",
       " 'pten': 803,\n",
       " 'highest': 804,\n",
       " 'inhibiting': 805,\n",
       " 'elements': 806,\n",
       " 'administered': 807,\n",
       " 'peptide': 808,\n",
       " 'since': 809,\n",
       " 'contribution': 810,\n",
       " 'diverse': 811,\n",
       " 'tumorigenesis': 812,\n",
       " 'laboratory': 813,\n",
       " 'shared': 814,\n",
       " 'structural': 815,\n",
       " 'preclinical': 816,\n",
       " 'radiation': 817,\n",
       " 'pcr': 818,\n",
       " 'progressive': 819,\n",
       " 'proposed': 820,\n",
       " 'context': 821,\n",
       " 'tyrosine': 822,\n",
       " 'subcutaneous': 823,\n",
       " 'agents': 824,\n",
       " 'identifying': 825,\n",
       " 'guidelines': 826,\n",
       " 'facilitate': 827,\n",
       " 'adhesion': 828,\n",
       " 'relationship': 829,\n",
       " 'rapid': 830,\n",
       " 'commonly': 831,\n",
       " 'hematopoietic': 832,\n",
       " 'standard': 833,\n",
       " 'find': 834,\n",
       " 'mainly': 835,\n",
       " 'downregulated': 836,\n",
       " 'assessment': 837,\n",
       " 'akt': 838,\n",
       " 'markedly': 839,\n",
       " 'examine': 840,\n",
       " 'chemotherapy': 841,\n",
       " 'culture': 842,\n",
       " 'published': 843,\n",
       " 'mutated': 844,\n",
       " 'patterns': 845,\n",
       " 'asthma': 846,\n",
       " 'imaging': 847,\n",
       " 'transgenic': 848,\n",
       " 'criteria': 849,\n",
       " 'trials': 850,\n",
       " 'prolonged': 851,\n",
       " 'properties': 852,\n",
       " 'made': 853,\n",
       " 'mild': 854,\n",
       " 'successful': 855,\n",
       " 'secretion': 856,\n",
       " 'body': 857,\n",
       " 'reverse': 858,\n",
       " 'point': 859,\n",
       " 'management': 860,\n",
       " 'toxicity': 861,\n",
       " 'exons': 862,\n",
       " 'change': 863,\n",
       " 'exposure': 864,\n",
       " 'metabolism': 865,\n",
       " 'activities': 866,\n",
       " 'months': 867,\n",
       " 'forms': 868,\n",
       " 'polymerase': 869,\n",
       " 'processes': 870,\n",
       " 'remained': 871,\n",
       " 'lymphocytes': 872,\n",
       " 'necessary': 873,\n",
       " 'differential': 874,\n",
       " 'allelic': 875,\n",
       " 'amino': 876,\n",
       " 'screened': 877,\n",
       " 'protection': 878,\n",
       " 'wide': 879,\n",
       " 'consent': 880,\n",
       " 'mean': 881,\n",
       " 'highlight': 882,\n",
       " 'muscular': 883,\n",
       " 'apoptotic': 884,\n",
       " 'events': 885,\n",
       " 'particular': 886,\n",
       " 'stage': 887,\n",
       " 'understand': 888,\n",
       " 'degradation': 889,\n",
       " 'fully': 890,\n",
       " 'validated': 891,\n",
       " 'capable': 892,\n",
       " 'retroviral': 893,\n",
       " 'genomics': 894,\n",
       " 'lupus': 895,\n",
       " 'hepatic': 896,\n",
       " 'design': 897,\n",
       " 'vaccination': 898,\n",
       " 'nine': 899,\n",
       " 'demonstrates': 900,\n",
       " 'green': 901,\n",
       " 'address': 902,\n",
       " 'member': 903,\n",
       " 'incidence': 904,\n",
       " 'showing': 905,\n",
       " 'global': 906,\n",
       " 'investigation': 907,\n",
       " 'mesenchymal': 908,\n",
       " 'suicide': 909,\n",
       " 'inhibitory': 910,\n",
       " 'along': 911,\n",
       " 'confirm': 912,\n",
       " 'mirnas': 913,\n",
       " 'systems': 914,\n",
       " 'promoting': 915,\n",
       " 'inducing': 916,\n",
       " 'upstream': 917,\n",
       " 'discovery': 918,\n",
       " 'finding': 919,\n",
       " 'certain': 920,\n",
       " 'symptoms': 921,\n",
       " 'splice': 922,\n",
       " 'unknown': 923,\n",
       " 'intron': 924,\n",
       " 'prevent': 925,\n",
       " 'day': 926,\n",
       " 'according': 927,\n",
       " 'counseling': 928,\n",
       " 'attenuated': 929,\n",
       " 'uptake': 930,\n",
       " 'marked': 931,\n",
       " 'versus': 932,\n",
       " 'serve': 933,\n",
       " 'components': 934,\n",
       " 'enhances': 935,\n",
       " 'powerful': 936,\n",
       " 'another': 937,\n",
       " 'causing': 938,\n",
       " 'ethnic': 939,\n",
       " 'technology': 940,\n",
       " 'become': 941,\n",
       " 'molecule': 942,\n",
       " 'activating': 943,\n",
       " 'treat': 944,\n",
       " 'interfering': 945,\n",
       " 'compound': 946,\n",
       " 'glucose': 947,\n",
       " 'sporadic': 948,\n",
       " 'pediatric': 949,\n",
       " 'rats': 950,\n",
       " 'bladder': 951,\n",
       " 'reports': 952,\n",
       " 'primarily': 953,\n",
       " 'cytokines': 954,\n",
       " 'origin': 955,\n",
       " 'mammalian': 956,\n",
       " 'sustained': 957,\n",
       " 'initiation': 958,\n",
       " 'larger': 959,\n",
       " 'generate': 960,\n",
       " 'needed': 961,\n",
       " 'transcripts': 962,\n",
       " 'displayed': 963,\n",
       " 'stability': 964,\n",
       " 'killing': 965,\n",
       " 'whole': 966,\n",
       " 'whose': 967,\n",
       " 'extensive': 968,\n",
       " 'prognostic': 969,\n",
       " 'insights': 970,\n",
       " 'transmission': 971,\n",
       " 'pulmonary': 972,\n",
       " 'males': 973,\n",
       " 'widely': 974,\n",
       " 'discuss': 975,\n",
       " 'average': 976,\n",
       " 'conserved': 977,\n",
       " 'substantial': 978,\n",
       " 'sensitive': 979,\n",
       " 'minor': 980,\n",
       " 'embryonic': 981,\n",
       " 'american': 982,\n",
       " 'cdna': 983,\n",
       " 'glioblastoma': 984,\n",
       " 'signals': 985,\n",
       " 'advances': 986,\n",
       " 'benefit': 987,\n",
       " 'clear': 988,\n",
       " 'cervical': 989,\n",
       " 'establish': 990,\n",
       " 'newly': 991,\n",
       " 'possibility': 992,\n",
       " 'occurred': 993,\n",
       " 'overcome': 994,\n",
       " 'allow': 995,\n",
       " 'best': 996,\n",
       " 'indicates': 997,\n",
       " 'numerous': 998,\n",
       " 'decision': 999,\n",
       " ...}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocabulary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert a given document into a vector of word counts\n",
    "def document_vector(document):\n",
    "    id = document[1]\n",
    "    counts = defaultdict(int)\n",
    "    for token in document[0]:\n",
    "        if token in vocabulary:\n",
    "            token_id = vocabulary[token]\n",
    "            counts[token_id] += 1\n",
    "    counts = sorted(counts.items())\n",
    "    keys = [x[0] for x in counts]\n",
    "    values = [x[1] for x in counts]\n",
    "    return (id, SparseVector(len(vocabulary), keys, values))\n",
    "\n",
    "# Process all the documents into word vector using above function\n",
    "documents = tokens.zipWithIndex().map(document_vector).map(list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[0,\n",
       "  SparseVector(29342, {1: 1.0, 2: 1.0, 3: 3.0, 5: 4.0, 8: 1.0, 13: 1.0, 24: 1.0, 34: 1.0, 35: 1.0, 38: 1.0, 39: 1.0, 51: 1.0, 53: 3.0, 54: 1.0, 59: 1.0, 63: 1.0, 64: 1.0, 92: 1.0, 105: 1.0, 134: 2.0, 151: 1.0, 156: 1.0, 171: 1.0, 179: 1.0, 183: 1.0, 198: 1.0, 222: 1.0, 256: 1.0, 280: 1.0, 330: 1.0, 363: 1.0, 370: 1.0, 391: 1.0, 418: 1.0, 434: 2.0, 448: 1.0, 450: 1.0, 459: 1.0, 505: 1.0, 513: 1.0, 521: 1.0, 529: 2.0, 734: 1.0, 755: 2.0, 790: 1.0, 833: 1.0, 907: 1.0, 965: 1.0, 1100: 1.0, 1109: 4.0, 1140: 2.0, 1180: 1.0, 1226: 1.0, 1263: 1.0, 1390: 1.0, 1439: 1.0, 1645: 1.0, 1820: 1.0, 1971: 1.0, 2341: 1.0, 2413: 1.0, 2533: 1.0, 2580: 1.0, 3121: 2.0, 3150: 1.0, 3349: 3.0, 3396: 1.0, 3804: 1.0, 4344: 1.0, 5124: 1.0, 5159: 1.0, 7235: 3.0, 12074: 1.0, 12668: 1.0, 13944: 2.0, 14360: 1.0, 27411: 1.0, 27412: 1.0})],\n",
       " [1,\n",
       "  SparseVector(29342, {4: 1.0, 5: 1.0, 10: 1.0, 23: 1.0, 25: 1.0, 43: 1.0, 291: 1.0, 298: 1.0, 394: 1.0, 595: 1.0, 1127: 1.0, 1275: 1.0, 1326: 2.0, 1783: 1.0, 1875: 1.0, 3314: 1.0, 3805: 1.0})],\n",
       " [2,\n",
       "  SparseVector(29342, {0: 1.0, 1: 1.0, 2: 1.0, 3: 1.0, 5: 3.0, 6: 1.0, 8: 1.0, 9: 1.0, 10: 1.0, 22: 1.0, 28: 1.0, 35: 1.0, 39: 3.0, 40: 1.0, 42: 1.0, 45: 1.0, 51: 1.0, 54: 1.0, 62: 1.0, 84: 6.0, 93: 1.0, 102: 1.0, 103: 1.0, 107: 1.0, 117: 1.0, 137: 1.0, 138: 2.0, 142: 1.0, 197: 1.0, 251: 1.0, 260: 1.0, 262: 2.0, 270: 2.0, 278: 1.0, 280: 1.0, 347: 1.0, 403: 1.0, 423: 1.0, 522: 1.0, 523: 1.0, 554: 1.0, 611: 1.0, 624: 1.0, 642: 3.0, 661: 1.0, 699: 1.0, 710: 1.0, 742: 1.0, 772: 1.0, 775: 1.0, 795: 1.0, 804: 1.0, 807: 1.0, 879: 1.0, 888: 1.0, 929: 1.0, 1034: 2.0, 1074: 1.0, 1105: 1.0, 1115: 1.0, 1152: 2.0, 1173: 1.0, 1379: 1.0, 1647: 1.0, 1713: 1.0, 1745: 1.0, 1765: 1.0, 1828: 1.0, 2157: 1.0, 2426: 1.0, 2601: 1.0, 2736: 1.0, 2811: 1.0, 2854: 1.0, 2905: 1.0, 3392: 2.0, 3484: 1.0, 3572: 1.0, 4243: 1.0, 4514: 1.0, 5030: 1.0, 5503: 1.0, 5864: 1.0, 6435: 1.0, 7935: 1.0})]]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "documents.take(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = documents.toDF(['doc_id', 'features'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------+--------------------+\n",
      "|doc_id|            features|\n",
      "+------+--------------------+\n",
      "|     0|(29342,[1,2,3,5,8...|\n",
      "|     1|(29342,[4,5,10,23...|\n",
      "|     2|(29342,[0,1,2,3,5...|\n",
      "|     3|(29342,[0,1,4,5,8...|\n",
      "|     4|(29342,[0,1,3,4,5...|\n",
      "|     5|(29342,[5,9,11,15...|\n",
      "|     6|(29342,[0,1,2,3,4...|\n",
      "|     7|(29342,[0,1,4,5,8...|\n",
      "|     8|(29342,[1,2,4,8,1...|\n",
      "|     9|(29342,[0,1,2,3,4...|\n",
      "|    10|(29342,[0,5,6,11,...|\n",
      "|    11|(29342,[2,3,4,5,8...|\n",
      "|    12|(29342,[0,1,3,4,5...|\n",
      "|    13|(29342,[0,2,3,4,5...|\n",
      "|    14|(29342,[1,3,5,8,1...|\n",
      "|    15|(29342,[0,1,2,15,...|\n",
      "|    16|(29342,[1,4,6,8,9...|\n",
      "|    17|(29342,[0,2,15,17...|\n",
      "|    18|(29342,[0,4,15,25...|\n",
      "|    19|(29342,[0,2,5,12,...|\n",
      "+------+--------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create spark dataframe\n",
    "#df = spark.createDataFrame(documents, [\"doc_id\", \"features\"])\n",
    "# split into training and testing data\n",
    "dfs = df.randomSplit([0.85, 0.15])\n",
    "dfTrain = dfs[0]\n",
    "dfTest = dfs[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------+--------------------+\n",
      "|doc_id|            features|\n",
      "+------+--------------------+\n",
      "|     0|(29342,[1,2,3,5,8...|\n",
      "|     1|(29342,[4,5,10,23...|\n",
      "|     2|(29342,[0,1,2,3,5...|\n",
      "|     3|(29342,[0,1,4,5,8...|\n",
      "|     4|(29342,[0,1,3,4,5...|\n",
      "|     5|(29342,[5,9,11,15...|\n",
      "|     6|(29342,[0,1,2,3,4...|\n",
      "|     7|(29342,[0,1,4,5,8...|\n",
      "|     8|(29342,[1,2,4,8,1...|\n",
      "|     9|(29342,[0,1,2,3,4...|\n",
      "|    10|(29342,[0,5,6,11,...|\n",
      "|    11|(29342,[2,3,4,5,8...|\n",
      "|    13|(29342,[0,2,3,4,5...|\n",
      "|    14|(29342,[1,3,5,8,1...|\n",
      "|    15|(29342,[0,1,2,15,...|\n",
      "|    16|(29342,[1,4,6,8,9...|\n",
      "|    17|(29342,[0,2,15,17...|\n",
      "|    18|(29342,[0,4,15,25...|\n",
      "|    19|(29342,[0,2,5,12,...|\n",
      "|    20|(29342,[0,1,2,7,8...|\n",
      "+------+--------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dfTrain.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11210"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfTrain.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Perplexity=8.855095541683063 for K=2, alpha=25.0, beta=0.05\n",
      "1285.8037214279175\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-34-82ae51885e5f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     20\u001b[0m         \u001b[0mlda_model\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetParams\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mK\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdocConcentration\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0malpha\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtopicConcentration\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbeta\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m         \u001b[0;31m# fit model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m         \u001b[0mdf_fit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlda_model\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdfTrain\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m         \u001b[0;31m# get fit perplexity\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m         \u001b[0mdf_perp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_fit\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlogPerplexity\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdfTest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/spark/python/pyspark/ml/base.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, dataset, params)\u001b[0m\n\u001b[1;32m    130\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_fit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    131\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 132\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_fit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    133\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    134\u001b[0m             raise ValueError(\"Params must be either a param map or a list/tuple of param maps, \"\n",
      "\u001b[0;32m/usr/local/spark/python/pyspark/ml/wrapper.py\u001b[0m in \u001b[0;36m_fit\u001b[0;34m(self, dataset)\u001b[0m\n\u001b[1;32m    286\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    287\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_fit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdataset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 288\u001b[0;31m         \u001b[0mjava_model\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_fit_java\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    289\u001b[0m         \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_create_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjava_model\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    290\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_copyValues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/spark/python/pyspark/ml/wrapper.py\u001b[0m in \u001b[0;36m_fit_java\u001b[0;34m(self, dataset)\u001b[0m\n\u001b[1;32m    283\u001b[0m         \"\"\"\n\u001b[1;32m    284\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_transfer_params_to_java\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 285\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_java_obj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    286\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    287\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_fit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdataset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1253\u001b[0m             \u001b[0mproto\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mEND_COMMAND_PART\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1254\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1255\u001b[0;31m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1256\u001b[0m         return_value = get_return_value(\n\u001b[1;32m   1257\u001b[0m             answer, self.gateway_client, self.target_id, self.name)\n",
      "\u001b[0;32m/usr/local/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36msend_command\u001b[0;34m(self, command, retry, binary)\u001b[0m\n\u001b[1;32m    983\u001b[0m         \u001b[0mconnection\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_connection\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    984\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 985\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconnection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    986\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mbinary\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    987\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_create_connection_guard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconnection\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36msend_command\u001b[0;34m(self, command)\u001b[0m\n\u001b[1;32m   1150\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1151\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1152\u001b[0;31m             \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msmart_decode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstream\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1153\u001b[0m             \u001b[0mlogger\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdebug\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Answer received: {0}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0manswer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1154\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0manswer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mproto\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mRETURN_MESSAGE\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/socket.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    584\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    585\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 586\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    587\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    588\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_timeout_occurred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# choose different values of topics\n",
    "Ks = [k for k in range(2, 21)]\n",
    "# topic prior, determines how sparse or non-sparse the topics are \n",
    "betas = [0.05, 0.1, 2., 5.]\n",
    "# initialize LDA model \n",
    "lda_model = LDA(seed=10, optimizer='online', maxIter=2000)\n",
    "# number of documents\n",
    "D = dfTrain.count()\n",
    "\n",
    "output = []\n",
    "\n",
    "for K in Ks:\n",
    "    for beta in betas:\n",
    "        t=time.time()\n",
    "        # per document proportion prior, based on Griffiths et. al, 2004\n",
    "        alpha = [50/K for _ in range(K)]\n",
    "        # set parameters in model accordingly\n",
    "        lda_model.setParams(k=K, docConcentration=alpha, topicConcentration=beta)\n",
    "        # fit model\n",
    "        df_fit = lda_model.fit(dfTrain)\n",
    "        # get fit perplexity\n",
    "        df_perp = df_fit.logPerplexity(dfTest)\n",
    "        \n",
    "        print('Perplexity={} for K={}, alpha={}, beta={}'.format(df_perp, K, 50/K, beta))\n",
    "        output.append(['Perplexity={} for K={}, alpha={}, beta={}'.format(df_perp, K, 50/K, beta)])\n",
    "        print(time.time()-t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('modelTuningResults', 'w') as f:\n",
    "    for item in output:\n",
    "        f.write(\"%s\\n\" % item)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
